platform/wayland/cog-platform-wl.c

```diff
@@ -281,22 +342,258 @@ static struct {
+    wl_shell_surface_set_toplevel(win_data.shell_context->wl_shell_data.shell_surface);
+
+    /* wl_shell needs an initial surface configuration. */
+    configure_surface_geometry(0, 0);
+}
+
+static const struct wl_shell_surface_listener shell_popup_surface_listener = {
+    .ping = shell_popup_surface_ping,
+    .configure = shell_popup_surface_configure,
+    .popup_done = shell_popup_surface_popup_done,
+};
+
+static void
+create_wl_shell_popup()
+{
+    popup_data.shell_surface = wl_shell_get_shell_surface(wl_data.shell, popup_data.wl_surface);
+    g_assert(popup_data.shell_surface);
+
+    wl_shell_surface_add_listener(popup_data.shell_surface, &shell_popup_surface_listener, NULL);
+    wl_shell_surface_set_popup(popup_data.shell_surface, wl_data.seat, wl_data.event_serial, win_data.wl_surface, 0,
+                               (win_data.height - popup_data.height), 0);
+
+    display_popup();
+}
+
+static void
+destroy_shell_surface()
+{
+    g_clear_pointer(&win_data.shell_context->wl_shell_data.shell_surface, wl_shell_surface_destroy);
+}
+
+wl_shell_data wayland_data = {.functions = {.enter_fullscreen = &wl_shell_enter_fullscreen,
+                                            .exit_fullscreen = &wl_shell_exit_fullscreen,
+                                            .destroy_shell = &destroy_wl_shell,
+                                            .maximize_surface = &maximize_wl_shell_surface,
+                                            .create_window = &create_wl_shell_window,
+                                            .create_popup = &create_wl_shell_popup,
+                                            .surface_destroy = &destroy_shell_surface}};
+
+// f shell operations
+static void
+f_shell_enter_fullscreen()
+{
+    win_data.should_resize_to_largest_output = true;
+    resize_to_largest_output();
+}
+static void
+f_shell_exit_fullscreen()
+{
+    configure_surface_geometry(win_data.width_before_fullscreen, win_data.height_before_fullscreen);
+    resize_window();
+}
+static void
+destroy_f_shell()
+{
+    zwp_fullscreen_shell_v1_destroy(wl_data.fshell);
+}
+static void
+create_f_shell_window()
+{
+    zwp_fullscreen_shell_v1_present_surface(wl_data.fshell,
+                                            win_data.wl_surface,
+                                            ZWP_FULLSCREEN_SHELL_V1_PRESENT_METHOD_DEFAULT,
+                                            NULL);
+
+    /* Configure the surface so that it respects the width and height
+        * environment variables */
+    configure_surface_geometry(0, 0);
+}
+
+f_shell_data f_data = {.functions = {.exit_fullscreen = &f_shell_exit_fullscreen,
+                                     .destroy_shell = &destroy_f_shell,
+                                     .create_window = &create_f_shell_window,
+                                     .enter_fullscreen = &f_shell_enter_fullscreen}};
+
+//default executed when no shell exist
+static void
```

Comment:
I would declare these before all the shell implementations, so that when shells are missing capabilities (like with f_shell above), you can reference these default implementations. I also note that there doesn't seem to be a full set of functions or an accompanying shell_functions struct for these.